server/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportResizeAction.java

```diff
@@ -129,12 +134,33 @@ static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(final Resi
         targetIndexSettingsBuilder.remove(IndexMetadata.SETTING_HISTORY_UUID);
         final Settings targetIndexSettings = targetIndexSettingsBuilder.build();
         final int numShards;
+        ByteSizeValue maxSinglePrimaryShardSize = resizeRequest.getMaxSinglePrimaryShardSize();
         if (IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.exists(targetIndexSettings)) {
+            if (resizeRequest.getResizeType() == ResizeType.SHRINK && maxSinglePrimaryShardSize != null) {
+                throw new IllegalArgumentException("Cannot set both index.number_of_shards and max_single_primary_shard_size" +
+                    " for the target index");
+            }
             numShards = IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.get(targetIndexSettings);
         } else {
             assert resizeRequest.getResizeType() != ResizeType.SPLIT : "split must specify the number of shards explicitly";
             if (resizeRequest.getResizeType() == ResizeType.SHRINK) {
-                numShards = 1;
+                if (maxSinglePrimaryShardSize != null) {
+                    int sourceIndexShardsNum = sourceMetadata.getNumberOfShards();
+                    long sourceIndexStorageBytes = indexStoreStats.getSizeInBytes();
+                    long maxSingleShardSizeBytes = maxSinglePrimaryShardSize.getBytes();
+                    long minShardsNum = sourceIndexStorageBytes / maxSingleShardSizeBytes;
+                    if (minShardsNum * maxSingleShardSizeBytes < sourceIndexStorageBytes) {
+                        minShardsNum = minShardsNum + 1;
+                    }
+                    if (minShardsNum > sourceIndexShardsNum) {
+                        throw new IllegalArgumentException("The target index's shards number[" + minShardsNum +
+                            "] is greater than the source index's shards number[" + sourceIndexShardsNum + "]");
+                    }
```

Comment:
I wonder if this should throw an error, or if it would be better to allow the shrink with the same number of shards as are currently available (basically `min(sourceIndexShardsNum, minShardsNum)`)

I am on the fence about this, since it can make the ILM usage almost useless if time-based rollover is used (imagine a period of high ingestion where an index with 2 primary shards has trouble initially rolling over, but finally does at 200gb, and then tries to shrink to `50gb` shards and gets stuck in an error forever because it's trying to "shrink" to 4 shards)

Reply:
That's a problem, I have set the target index's shards number to `sourceIndexShardsNum` if `minShardsNum` is greater than `sourceIndexShardsNum`, I have also added a log in the code and some description in the document about this.